### PR TITLE
fix configuration unsubscribe route

### PR DESCRIPTION
**Pull request by xuybin**
fix configuration unsubscribe route

# Description

when  "PSOT http://localhost:daprPort/v1.0-alpha1/configuration/subscription-id/unsubscribe"  report an error:

## Issue reference

goroutine 15 [running]:
github.com/dapr/dapr/pkg/http.(*api).getStateStoreName(...)
        D:/a/dapr/dapr/pkg/http/api.go:1358
github.com/dapr/dapr/pkg/http.(*api).getConfigurationStoreWithRequestValidation(0xc0010bc3c0, 0xc00102c000)
        D:/a/dapr/dapr/pkg/http/api.go:768 +0x4e5
github.com/dapr/dapr/pkg/http.(*api).onUnsubscribeConfiguration(0xc0010bc3c0, 0xc00102c000)
        D:/a/dapr/dapr/pkg/http/api.go:984 +0x3e
github.com/dapr/dapr/pkg/http.(*server).unescapeRequestParametersHandler.func1(0xc00102c000)
        D:/a/dapr/dapr/pkg/http/server.go:303 +0x71
github.com/fasthttp/router.(*Router).Handler(0xc0010ca460, 0xc00102c000)
        C:/Users/runneradmin/go/pkg/mod/github.com/fasthttp/router@v1.3.8/router.go:414 +0x46a
github.com/dapr/dapr/pkg/diagnostics.(*httpMetrics).FastHTTPMiddleware.func1(0xc00102c000)
        D:/a/dapr/dapr/pkg/diagnostics/http_monitoring.go:195 +0x13b
github.com/dapr/dapr/pkg/diagnostics.HTTPTraceMiddleware.func1(0xc00102c000)
        D:/a/dapr/dapr/pkg/diagnostics/http_tracing.go:56 +0xef
github.com/valyala/fasthttp.(*Server).serveConn(0xc00109e400, {0x568db48?, 0xc0003a21c0})
        C:/Users/runneradmin/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/server.go:2293 +0x1268
github.com/valyala/fasthttp.(*workerPool).workerFunc(0xc000f12780, 0xc000599a60)
        C:/Users/runneradmin/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/workerpool.go:223 +0xa9
github.com/valyala/fasthttp.(*workerPool).getCh.func1()
        C:/Users/runneradmin/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/workerpool.go:195 +0x38
created by github.com/valyala/fasthttp.(*workerPool).getCh
        C:/Users/runneradmin/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/workerpool.go:194 +0x1b0
The daprd process exited with error code: exit status 2

terminated signal received: shutting down
Error exiting Dapr: exit status 2

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing

### CODE REVIEW

File: pkg/http/api.go [hunk 1]
```diff
@@ -437,7 +437,7 @@ func (a *api) constructConfigurationEndpoints() []Endpoint {
 		},
 		{
 			Methods: []string{fasthttp.MethodPost},
-			Route:   "configuration/{configurationSubscribeID}/unsubscribe",
+			Route:   "configuration/{storeName}/{configurationSubscribeID}/unsubscribe",
```
Comment (daixiang0) [inline pkg/http/api.go:440]: Could we have a test for this?

Comment (xuybin) [reply pkg/http/api.go:440]: I don't see any existing tests for the "https://v1-8.docs.dapr.io/reference/api/configuration_api/"

Comment (daixiang0) [reply pkg/http/api.go:440]: Please add it.

Comment (xuybin) [reply pkg/http/api.go:440]: func TestV1Alpha1ConfigurationUnsubscribe(t *testing.T) {
	fakeServer := newFakeHTTPServer()

	var fakeConfigurationStore configuration.Store = &fakeConfigurationStore{}

	storeName := "store1"
	badStoreName := "nonExistStore"

	fakeConfigurationStores := map[string]configuration.Store{
		storeName: fakeConfigurationStore,
	}
	testAPI := &api{
		resiliency:          resiliency.New(nil),
		configurationStores: fakeConfigurationStores,
	}
	fakeServer.StartServer(testAPI.constructConfigurationEndpoints())

	t.Run("subscribe and unsubscribe configurations", func(t *testing.T) {
		
		apiPath1 := fmt.Sprintf("v1.0-alpha1/configuration/%s/subscribe", storeName)
		resp1 := fakeServer.DoRequest("GET", apiPath1, nil, nil)
		assert.Equal(t, 200, resp1.StatusCode, "subscribe configuration store, should return 200")


		rspMap1 := resp1.JSONBody.(interface{})
		assert.NotNil(t, rspMap1)
		//rspMap1.(map[string]interface{})["id"].(string)
		
		apiPath2 := fmt.Sprintf("v1.0-alpha1/configuration/%s/%s/unsubscribe", storeName,rspMap1.(map[string]interface{})["id"].(string))
		
		resp2 := fakeServer.DoRequest("POST", apiPath2, nil, nil)

		assert.Equal(t, 204, resp2.StatusCode, "unsubscribe configuration store,should return 204")
	})
}

Comment (xuybin) [reply pkg/http/api.go:440]:
```suggestion
  like this at /pkg/http/api_test.go
  func TestV1Alpha1ConfigurationUnsubscribe(t *testing.T) {
	fakeServer := newFakeHTTPServer()

	var fakeConfigurationStore configuration.Store = &fakeConfigurationStore{}

	storeName := "store1"

	fakeConfigurationStores := map[string]configuration.Store{
		storeName: fakeConfigurationStore,
	}
	testAPI := &api{
		resiliency:          resiliency.New(nil),
		configurationStores: fakeConfigurationStores,
	}
	fakeServer.StartServer(testAPI.constructConfigurationEndpoints())

	t.Run("subscribe and unsubscribe configurations", func(t *testing.T) {
		
		apiPath1 := fmt.Sprintf("v1.0-alpha1/configuration/%s/subscribe", storeName)
		resp1 := fakeServer.DoRequest("GET", apiPath1, nil, nil)
		assert.Equal(t, 200, resp1.StatusCode, "subscribe configuration store, should return 200")


		rspMap1 := resp1.JSONBody.(interface{})
		assert.NotNil(t, rspMap1)
		//rspMap1.(map[string]interface{})["id"].(string)
		
		apiPath2 := fmt.Sprintf("v1.0-alpha1/configuration/%s/%s/unsubscribe", storeName,rspMap1.(map[string]interface{})["id"].(string))
		
		resp2 := fakeServer.DoRequest("POST", apiPath2, nil, nil)

		assert.Equal(t, 204, resp2.StatusCode, "unsubscribe configuration store,should return 204")
	})
}
```

**Review by xuybin — COMMENTED** [review]
add test for subscribe and unsubscribe configurations

**Review by xuybin — COMMENTED** [review]
style clean up comments

File: pkg/http/api.go (hunk 1 above)
Comment (xuybin) [reply pkg/http/api.go:440]: https://v1-8.docs.dapr.io/reference/api/configuration_api/#unsubscribe-configuration describes,unsubscribe action is GET, but here is POST, is this normal?
`Methods: []string{fasthttp.MethodPost},
			Route:   "configuration/{configurationSubscribeID}/unsubscribe",`

**Comment by yaron2** [comment]
cc @pravinpushkar

**Comment by pravinpushkar** [comment]
@yaron2 @xuybin 
The issue that I am facing is related to **Subscribe** configuration - https://github.com/dapr/dapr/issues/4968
Is that also known issue and is this PR going to handle that ?

**Comment by xuybin** [comment]
> 



> @yaron2 @xuybin 我面临的问题与**订阅**配置有关 - #4968 这也是已知问题吗，此 PR 将如何处理此问题？
start dapr sidecar  '--app-port' parameter must be provided,
and the app corresponding to app-port must be started before dapr sidecar

File: pkg/http/api.go (hunk 1 above)
Comment (yaron2) [reply pkg/http/api.go:440]: No, this should be `GET`. Good catch! Please change the verb.

Comment (xuybin) [reply pkg/http/api.go:440]: > 不，这应该是 。抓得好！请更改动词。`GET`

 done

**Comment by daixiang0** [comment]
@xuybin please fix DCO issue.

**Comment by xuybin** [comment]
> **DCO ** — DCO

@daixiang0  done

**Comment by pravinpushkar** [comment]
General question  - do we have E2E tests for configuration APIs?

**Comment by yaron2** [comment]
> General question - do we have E2E tests for configuration APIs?

No, we'll need to add those in order to promote it to beta and stable.

**Review by yaron2 — DISMISSED** [review]

**Comment by yaron2** [comment]
@xuybin please fix the linter errors: https://github.com/dapr/dapr/runs/7728862124?check_suite_focus=true and we can then merge.

**Comment by xuybin** [comment]
> @xuybin please fix the linter errors: https://github.com/dapr/dapr/runs/7728862124?check_suite_focus=true and we can then merge.
@yaron2   done
fix lint error with command:
gofumpt -l -w ./pkg/http/api.go
gofumpt -l -w ./pkg/http/api_test.go

**Comment by yaron2** [comment]
> > @xuybin please fix the linter errors: https://github.com/dapr/dapr/runs/7728862124?check_suite_focus=true and we can then merge.
> > @yaron2   done
> > fix lint error with command:
> > gofumpt -l -w ./pkg/http/api.go
> > gofumpt -l -w ./pkg/http/api_test.go

One more: https://github.com/dapr/dapr/pull/4964/files

**Comment by xuybin** [comment]
> > > @xuybin please fix the linter errors: https://github.com/dapr/dapr/runs/7728862124?check_suite_focus=true and we can then merge.
> > > @yaron2   done
> > > fix lint error with command:
> > > gofumpt -l -w ./pkg/http/api.go
> > > gofumpt -l -w ./pkg/http/api_test.go
> 
> One more: https://github.com/dapr/dapr/pull/4964/files

done

**Review by yaron2 — APPROVED** [review]

**Pull request merged**